packages/webpack-extensive-lodash-replacement-plugin/index.js

```diff
@@ -332,74 +330,106 @@ const LODASH_MODULE_NAMES = [
+			if ( error ) {
+				reject(
+					createError(
+						`Could not find module ${ packageName } for import on ${ rootPath }.`,
+						error
+					)
+				);
+			}
+
+			resolve( context );
+		} );
+	} );
+}
+
 class ExtensiveLodashReplacementPlugin {
-	constructor( baseFile = './package.json' ) {
-		this.baseFile = path.resolve( baseFile );
-		let data;
+	constructor( baseDir = '.' ) {
+		this.baseDir = path.resolve( baseDir );
+	}
 
+	async initBaseLodashData() {
 		try {
-			data = fs.readFileSync( this.baseFile );
+			this.baseLodashContext = await getModuleForPath(
+				this.moduleResolver,
+				this.baseDir,
+				'lodash'
+			);
 		} catch ( error ) {
-			throwError( 'Could not read root package.json.', error );
+			throw createError( 'Could not find root `lodash`.' );
 		}
 
 		try {
-			const packageJson = JSON.parse( data );
-			this.lodashVersion =
-				packageJson && packageJson.dependencies && packageJson.dependencies[ 'lodash-es' ];
+			const baseLodashES = await getModuleForPath( this.moduleResolver, this.baseDir, 'lodash-es' );
+			this.baseLodashESVersion =
+				baseLodashES &&
+				baseLodashES.descriptionFileData &&
+				baseLodashES.descriptionFileData.version;
 		} catch ( error ) {
-			throwError( 'Could not parse root package.json.', error );
-		}
-
-		if ( ! this.lodashVersion ) {
-			throwError( 'No lodash-es dependency in root package.json.' );
+			throw createError( 'Could not find root `lodash-es`.' );
 		}
 
-		if ( ! semver.valid( this.lodashVersion ) ) {
-			throwError( `Invalid root package.json lodash-es version: ${ this.lodashVersion }.` );
+		if ( ! this.baseLodashESVersion ) {
+			throw createError( 'Could not determine root `lodash-es` version.' );
 		}
 	}
 
-	// Figure out the requested Lodash version range for a given import.
-	// It looks at the file with the import and traverses upwards, until it finds a package.json file
-	// with the requested dependency. It returns this range.
-	findRequestedLodashRange( file, packageName ) {
-		const finder = findPackageJson( path.dirname( file ) );
-		let version;
-
-		try {
-			while ( ! version ) {
-				const found = finder.next();
-
-				if ( found.filename === this.baseFile ) {
-					return this.lodashVersion;
-				}
+	// Figure out the version for a given import.
+	// It follows the node resolution algorithm until it finds the package, returning its version.
+	async findRequestedVersion( file, packageName ) {
+		const foundContext = await getModuleForPath(
+			this.moduleResolver,
+			path.dirname( file ),
+			packageName
+		);
+		const baseLodashContext = this.baseLodashContext;
 
-				const contents = found && found.value;
-				version = contents && contents.dependencies && contents.dependencies[ packageName ];
-			}
-		} catch ( error ) {
-			throwError( `Could not find requested range for import on ${ file }.`, error );
+		if ( foundContext.path === baseLodashContext.path ) {
+			return this.baseLodashESVersion;
```

Comment:
I don't understand this special case for `lodash` that's installed in the top-level `node_modules`. Why do we return `this.baseLodashESVersion` instead of `foundContext.descriptionFileData.version`?

With this special case, it's like the root `lodash` version doesn't matter at all and is ignored?

Reply:
That's an early exit that checks whether we resolved to the `lodash` in `/node_modules`. That means that we're resolving something in Calypso and not one of its dependencies.

In this situation, we don't need to check any versions if we keep `lodash` and `lodash-es` in sync in our `package.json`; we can just replace one for the other and call it a day.

Reply:
If we remove this check though, the version found in `foundResolveData` is exactly the same, isn't it? We can remove the extra condition.

I would add one more early check to init: verify that the `lodash` and `lodash-es` versions are identical in the top-level `node_modules`. And exit with error if they are not.

If we do this, we don't need to set `this.baseLodashResolveData`. It would be just a local variable in `initBaseLodashData`.

Reply:
Done!